packages/ckb-indexer/src/ckbIndexerFilter.ts

```diff
@@ -0,0 +1,341 @@
+  HexadecimalRange,
+  QueryOptions,
+  Script,
+  ScriptWrapper,
+  blockchain,
+} from "@ckb-lumos/base";
+import { SearchKey } from "./type";
+import { bytes } from "@ckb-lumos/codec";
+import { BI } from "@ckb-lumos/bi";
+import { decodeCursor } from "./indexerCursor";
+
+function convertQueryOptionToSearchKey(queryOptions: QueryOptions): SearchKey {
+  let searchKeyLock: Script | undefined;
+  let searchKeyType: Script | undefined;
+  let searchKey: SearchKey;
+
+  const queryLock = queryOptions.lock;
+  const queryType = queryOptions.type;
+  if (queryLock) {
+    if (instanceOfScriptWrapper(queryLock)) {
+      searchKeyLock = queryLock.script;
+    } else {
+      searchKeyLock = queryLock;
+    }
+  }
+  if (queryType && queryType !== "empty") {
+    if (instanceOfScriptWrapper(queryType)) {
+      searchKeyType = queryType.script;
+    } else {
+      searchKeyType = queryType;
+    }
+  }
+
+  if (searchKeyLock) {
+    searchKey = {
+      script: searchKeyLock,
+      scriptType: "lock",
+      scriptSearchMode: instanceOfScriptWrapper(queryLock)
+        ? queryLock.searchMode
+        : "prefix",
+      filter: {
+        script: searchKeyType,
+      },
+    };
+  } else if (searchKeyType) {
+    searchKey = {
+      script: searchKeyType,
+      scriptType: "type",
+      scriptSearchMode: instanceOfScriptWrapper(queryLock)
+        ? queryLock.searchMode
+        : "prefix",
+      filter: {
+        script: searchKeyLock,
+      },
+    };
+  } else {
+    throw new Error("query.lock and query.type can't be both empty");
+  }
+
+  if (queryOptions.fromBlock || queryOptions.toBlock) {
+    searchKey.filter!.blockRange = [
+      queryOptions.fromBlock || "0x0",
+      queryOptions.toBlock
+        ? BI.from(queryOptions.toBlock).add(1).toHexString()
+        : "0x" + Number.MAX_SAFE_INTEGER.toString(16),
+    ];
+  }
+
+  return searchKey;
+}
+
+function filterByQueryOptions(cells: Cell[], options: QueryOptions): Cell[] {
+  const searchKey = convertQueryOptionToSearchKey(options);
+  let filteredCells = cells.filter((cell) => filterBy(cell, searchKey));
+
+  if (options.argsLen && options.argsLen !== "any") {
+    filteredCells = filteredCells.filter(
+      (cell) =>
+        bytes.bytify(cell.cellOutput.lock.args).length === options.argsLen
+    );
+  }
+
+  if (!!options.data && options.data !== "any") {
+    if (
+      instanceOfDataWithSearchMode(options.data) &&
+      options.data.searchMode === "exact"
+    ) {
+      const dataSearch = options.data as DataWithSearchMode;
+      filteredCells = filteredCells.filter((cell) =>
+        bytes.equal(bytes.bytify(cell.data), bytes.bytify(dataSearch.data))
+      );
+    } else if (
+      instanceOfDataWithSearchMode(options.data) &&
+      options.data.searchMode === "prefix"
+    ) {
+      const dataSearch = options.data as DataWithSearchMode;
+      filteredCells = filteredCells.filter(
+        (cell) =>
+          Buffer.from(bytes.bytify(cell.data)).indexOf(
```

Comment:
We should avoid using the NodeJS API as possible as we can

```ts
const expectPrefix = bytes.bytify(dataSearch.data);
const actualPrefix = cell.data.slice(0, expectPrefix.length);
return bytes.equal(expectPrefix, actualPrefix);
```

Reply:
removed all `Buffer`